File: src/cpp/CalculadorIntegralPool.cpp
#include "CalculadorIntegralPool.h"
#include "ThreadPool.h"
#include "Funcion.h"
#include <thread>
#include <vector>
#include <numeric>

CalculadorIntegralPool::CalculadorIntegralPool(long double a, long double b, long long n, ThreadPool &pool)
    : a(a), b(b), n(n), pool(pool)
{
    if (n > 0)
    {
        h = (b - a) / n;
    }
    else
    {
        h = 0;
    }
}

/**
void CalculadorIntegral::worker(long long inicio, long long fin, long double &resultado)
{
    long double suma_local = 0.0L;
    for (long long i = inicio; i < fin; ++i)
    {
        suma_local += f(a + i * h);
    }
    resultado = suma_local;
}
*/

long double CalculadorIntegralPool::calcular()
{
    if (n == 0)
        return 0;

    long double area = (f(a) + f(b)) / 2.0L;

    int num_hilos = pool.getNum_hilos();

    long long trapecios_por_hilo = (n - 1) / num_hilos;

    std::vector<std::future<long double>> resultados_futuros;

    for (int i = 0; i < num_hilos; ++i)
    {
        long long inicio = 1 + i * trapecios_por_hilo;
        long long fin = (i == num_hilos - 1) ? n : inicio + trapecios_por_hilo;

        // Envolver la tarea en una lambda para capturar los parámetros por valor
        // Nota: antes era la funcion worker, ahora es una lambda
        resultados_futuros.emplace_back(
            pool.enqueue([this, inicio, fin]
                         {
                long double suma_local = 0.0L;
                for (long long j = inicio; j < fin; ++j) 
                {
                    suma_local += f(a + j * h);
                }
            return suma_local; }));
    }

    long double suma_total_parcial = 0.0L;

    // Recoger los resultados de las tareas
    for (auto &futuro : resultados_futuros)
    {
        suma_total_parcial += futuro.get();
    }

    area += suma_total_parcial;

    return area * h;
}

File: src/java/CalculadorIntegralPool.java
import java.math.BigDecimal;
import java.math.MathContext;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Future;

public class CalculadorIntegralPool {
    private final BigDecimal a, b, h;
    private final long n;
    private final ThreadPool pool;
    private static final MathContext MC = new MathContext(100); // Mayor precisión
    
    public CalculadorIntegralPool(BigDecimal a, BigDecimal b, long n, ThreadPool pool) {
        this.a = a;
        this.b = b;
        this.n = n;
        this.pool = pool;
        
        if (n > 0) {
            BigDecimal nBig = new BigDecimal(n);
            this.h = b.subtract(a, MC).divide(nBig, MC);
        } else {
            this.h = BigDecimal.ZERO;
        }
    }
    
    public BigDecimal calcular() {
        if (n == 0) return BigDecimal.ZERO;
        
        BigDecimal fa = Funcion.f(a);
        BigDecimal fb = Funcion.f(b);
        BigDecimal area = fa.add(fb, MC).divide(new BigDecimal("2"), MC);
        
        int numHilos = pool.getNumHilos();
        long trapezoidesPorHilo = (n - 1) / numHilos;
        List<Future<BigDecimal>> futures = new ArrayList<>();
        
        for (int i = 0; i < numHilos; i++) {
            final long inicio = 1 + i * trapezoidesPorHilo;
            final long fin = (i == numHilos - 1) ? n : inicio + trapezoidesPorHilo;
            
            futures.add(pool.enqueue(() -> {
                BigDecimal sumaLocal = BigDecimal.ZERO;
                for (long j = inicio; j < fin; j++) {
                    BigDecimal x = a.add(h.multiply(new BigDecimal(j), MC), MC);
                    sumaLocal = sumaLocal.add(Funcion.f(x), MC);
                }
                return sumaLocal;
            }));
        }
        
        // Recoger resultados
        for (Future<BigDecimal> future : futures) {
            try {
                area = area.add(future.get(), MC);
            } catch (Exception e) {
                throw new RuntimeException("Error en hilo del pool", e);
            }
        }
        
        return area.multiply(h, MC);
    }
}